platformio/platform-raspberrypi
# Arduino-Pico (Earlephilhower) support, PicoProbe Debugging

Massive upgrade.

* Add support for building for https://github.com/earlephilhower/arduino-pico
* As in the style of platform-ststm32, boards that can be used with both cores can be switched to the other core using `board_build.core = earlephilhower`
* Integrates with my already merged builder script at https://github.com/earlephilhower/arduino-pico/blob/master/tools/platformio-build.py
* Added all Arduino-Pico boards (Earle auto-gen'ed them through https://github.com/earlephilhower/arduino-pico/blob/master/tools/makeboards.py)
* Arduino core must use toolchain and tools from https://github.com/earlephilhower/pico-quick-toolchain/releases, which are finally in the registry
  * **no more**  `platform_packages` injection `platformio.ini` needed, all tools (except framework) sourced from registry   
* Replace outdated `platformio/tool-openocd-raspberrypi` (1 year old, before picoprobe merge) with `earlephilhower/tool-openocd-rp2040-earlephilhower` -- works for all frameworks and cores (also the mbed-os core)
* Added support for LittleFS filesystem building + uploading through picotool and OpenOCD (picoprobe etc.); Arduino-Pico core supports LittleFS
* Added and tested Picoprobe support for both ArduinoCore-mbed and Arduino-Pico, works great
* Reduced default debugging and OpenOCD upload speed to 1000kHz, the 5000kHz was giving me verification errors
* Add Arduino-Pico examples to CI
* Reprogrammed `upload_protocol = mbed` to make more sense, it previously tried to upload the `.bin` file to the drive instead of the `.uf2` file
* Update SVD file with latest from https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2040/hardware_regs/rp2040.svd

![grafik](https://user-images.githubusercontent.com/26485477/172937024-f8af8c4a-edab-4374-ae4b-a4a96db09e1e.png)

Test at will with 
```ini
[env]
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
framework = arduino
board_build.core = earlephilhower
board_build.filesystem_size = 0.5m

[env:pico]
board = pico
; if using picoprobe SWD upload / debugging
;upload_protocol = picoprobe
;debug_tool = picoprobe
; note: newer PicoProbe firmwares emulate a CMSIS-DAP, so you
; use "cmsis-dap" as upload_protocol and debug_tool.
; for more info see docs https://arduino-pico.readthedocs.io/en/latest/platformio.html
```

ToDo (all done at the moment of writing):
* create PR for platformio-docs, previous docs at https://arduino-pico.readthedocs.io/en/latest/platformio.html are outdated
* create PR for arduino-pico docs as docs is now outdated
* fix https://github.com/earlephilhower/arduino-pico/issues/612
* upload / make Earle upload `framework-arduinopico` package (after fix of course)
